client/asset/btc/electrum_client.go

```diff
@@ -0,0 +1,1298 @@
+	if err != nil {
+		return time.Time{}, err
+	}
+
+	if hdrsRes.Count != medianTimeBlocks {
+		ew.log.Warnf("Failed to retrieve headers for %d blocks since block %v, got %d",
+			medianTimeBlocks, height, hdrsRes.Count)
+	}
+	if hdrsRes.Count == 0 {
+		return time.Time{}, errors.New("no headers retrieved")
+	}
+
+	hdrReader := hex.NewDecoder(strings.NewReader(hdrsRes.HexConcat))
+
+	timestamps := make([]int64, 0, hdrsRes.Count)
+	for i := int64(0); i < int64(hdrsRes.Count); i++ {
+		hdr := &wire.BlockHeader{}
+		err = hdr.Deserialize(hdrReader)
+		if err != nil {
+			if i > 0 {
+				ew.log.Errorf("Failed to deserialize header for block %d: %v",
+					startHeight+i, err)
+				break // we have at least one time stamp, work with it
+			}
+			return time.Time{}, err
+		}
+		timestamps = append(timestamps, hdr.Timestamp.Unix())
+	}
+	// Naive way fetching each header separately, if we needed to use
+	// btc.calcMedianTime as a chainStamper:
+	// for i := height; i > height-medianTimeBlocks && i > 0; i-- {
+	// 	hdr, err := ew.getBlockHeaderByHeight(height)
+	// 	if err != nil {
+	// 		return time.Time{}, err
+	// 	}
+	// 	timestamps = append(timestamps, hdr.Timestamp.Unix())
+	// }
+
+	sort.Slice(timestamps, func(i, j int) bool {
+		return timestamps[i] < timestamps[j]
+	})
+
+	medianTimestamp := timestamps[len(timestamps)/2]
+	return time.Unix(medianTimestamp, 0), nil
+}
+
+func (ew *electrumWallet) getBlockHash(height int64) (*chainhash.Hash, error) {
+	hdr, err := ew.getBlockHeaderByHeight(height)
+	if err != nil {
+		return nil, err
+	}
+	hash := hdr.BlockHash()
+	return &hash, nil
+}
+
+func (ew *electrumWallet) getBestBlockHash() (*chainhash.Hash, error) {
+	inf, err := ew.wallet.GetInfo()
+	if err != nil {
+		return nil, err
+	}
+	return ew.getBlockHash(inf.SyncHeight)
+}
+
+func (ew *electrumWallet) getBestBlockHeight() (int32, error) {
+	inf, err := ew.wallet.GetInfo()
+	if err != nil {
+		return 0, err
+	}
+	return int32(inf.SyncHeight), nil
+}
+
+func (ew *electrumWallet) getBestBlockHeader() (*blockHeader, error) {
+	inf, err := ew.wallet.GetInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	hdr, err := ew.getBlockHeaderByHeight(inf.SyncHeight)
+	if err != nil {
+		return nil, err
+	}
+
+	header := &blockHeader{
+		Hash:              hdr.BlockHash().String(),
+		Height:            inf.SyncHeight,
+		Confirmations:     1, // it's the head
+		Time:              hdr.Timestamp.Unix(),
+		PreviousBlockHash: hdr.PrevBlock.String(),
+	}
+	return header, nil
+}
+
+func (ew *electrumWallet) balances() (*GetBalancesResult, error) {
+	eBal, err := ew.wallet.GetBalance()
+	if err != nil {
+		return nil, err
+	}
+	// NOTE: Nothing from the Electrum wallet's response indicates trusted vs.
+	// untrusted. To allow unconfirmed coins to be spent, we treat both
+	// confirmed and unconfirmed as trusted. This is like dogecoind's handling
+	// of balance. TODO: listunspent -> checkWalletTx(txid) -> for each
+	// input, checkWalletTx(prevout) and ismine(addr)
```

Comment:
> TODO: listunspent -> checkWalletTx(txid) -> for each input, checkWalletTx(prevout) and ismine(addr)

Seems like you're checking that we actually own the utxos?

Reply:
Right, this TODO is alluding to an "ownsInputs" kind of function like we have for `spvWallet`.  In this iteration it's just like DOGE where there's no distinction between trusted vs. untrusted.  I definitely want to try this before release.